File: tests/wasm/test_range_filter.js
/**
 * WASM binding tests for RANGE_FILTER indicator.
 * These tests mirror the Rust unit tests to ensure WASM bindings work correctly.
 */
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { dirname } from 'path';
import { 
    loadTestData, 
    assertArrayClose, 
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS 
} from './test_utils.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

let wasm;
let testData;

test.before(async () => {
    
    try {
        const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
        const importPath = process.platform === 'win32' 
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);
        
    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }
    
    testData = loadTestData();
});

test('RANGE_FILTER accuracy', () => {
    
    const close = new Float64Array(testData.close);
    
    
    const result = wasm.range_filter_js(close);
    
    assert(result.filter, 'Should have filter output');
    assert(result.high_band, 'Should have high_band output');
    assert(result.low_band, 'Should have low_band output');
    assert.strictEqual(result.filter.length, close.length);
    assert.strictEqual(result.high_band.length, close.length);
    assert.strictEqual(result.low_band.length, close.length);
    
    
    const expectedFilter = [
        59589.73987817684, 
        59589.73987817684, 
        59589.73987817684, 
        59589.73987817684, 
        59589.73987817684
    ];
    
    const expectedHigh = [
        60935.63924911415,
        60906.58379951138,
        60874.2002431993,
        60838.79850154794,
        60810.879398758305
    ];
    
    const expectedLow = [
        58243.84050723953,
        58272.8959568423,
        58305.27951315438,
        58340.68125480574,
        58368.60035759538
    ];
    
    const tolerance = 1e-10; 
    
    
    const last5Filter = result.filter.slice(-5);
    for (let i = 0; i < 5; i++) {
        const diff = Math.abs(last5Filter[i] - expectedFilter[i]);
        assert(
            diff < tolerance,
            `Filter[${i}] mismatch: expected ${expectedFilter[i]}, got ${last5Filter[i]} (diff: ${diff})`
        );
    }
    
    
    const last5High = result.high_band.slice(-5);
    for (let i = 0; i < 5; i++) {
        const diff = Math.abs(last5High[i] - expectedHigh[i]);
        assert(
            diff < tolerance,
            `High Band[${i}] mismatch: expected ${expectedHigh[i]}, got ${last5High[i]} (diff: ${diff})`
        );
    }
    
    
    const last5Low = result.low_band.slice(-5);
    for (let i = 0; i < 5; i++) {
        const diff = Math.abs(last5Low[i] - expectedLow[i]);
        assert(
            diff < tolerance,
            `Low Band[${i}] mismatch: expected ${expectedLow[i]}, got ${last5Low[i]} (diff: ${diff})`
        );
    }
});

test('RANGE_FILTER default candles', () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.range_filter_js(close);
    
    assert(result.filter, 'Should have filter output');
    assert(result.high_band, 'Should have high_band output');
    assert(result.low_band, 'Should have low_band output');
    assert.strictEqual(result.filter.length, close.length);
    assert.strictEqual(result.high_band.length, close.length);
    assert.strictEqual(result.low_band.length, close.length);
});

test('RANGE_FILTER partial params', () => {
    
    const close = new Float64Array(testData.close);
    
    
    let result = wasm.range_filter_js(close, 2.5);
    assert.strictEqual(result.filter.length, close.length);
    
    
    result = wasm.range_filter_js(close, 2.5, 15);
    assert.strictEqual(result.filter.length, close.length);
    
    
    result = wasm.range_filter_js(close, 2.5, 15, true, 20);
    assert.strictEqual(result.filter.length, close.length);
});

test('RANGE_FILTER empty input', () => {
    
    const empty = new Float64Array([]);
    
    assert.throws(() => {
        wasm.range_filter_js(empty);
    }, /Input data slice is empty/);
});

test('RANGE_FILTER all NaN', () => {
    
    const allNaN = new Float64Array(100);
    allNaN.fill(NaN);
    
    assert.throws(() => {
        wasm.range_filter_js(allNaN);
    }, /All values are NaN/);
});

test('RANGE_FILTER invalid period', () => {
    
    const data = new Float64Array([1.0, 2.0, 3.0]);
    
    
    assert.throws(() => {
        wasm.range_filter_js(data, 2.618, 10);
    }, /Invalid period/);
    
    
    assert.throws(() => {
        wasm.range_filter_js(data, 2.618, 0);
    }, /Invalid period|Invalid range_period/);
});

test('RANGE_FILTER invalid range_size', () => {
    
    const data = new Float64Array([1.0, 2.0, 3.0, 4.0, 5.0]);
    
    
    assert.throws(() => {
        wasm.range_filter_js(data, 0.0);
    }, /Invalid range_size/);
    
    
    assert.throws(() => {
        wasm.range_filter_js(data, -1.0);
    }, /Invalid range_size/);
    
    
    assert.throws(() => {
        wasm.range_filter_js(data, NaN);
    }, /Invalid range_size/);
});

test('RANGE_FILTER NaN handling', () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.range_filter_js(close);
    
    
    
    if (result.filter.length > 50) {
        for (let i = 50; i < result.filter.length; i++) {
            assert(!isNaN(result.filter[i]), `Found unexpected NaN in filter at index ${i}`);
            assert(!isNaN(result.high_band[i]), `Found unexpected NaN in high_band at index ${i}`);
            assert(!isNaN(result.low_band[i]), `Found unexpected NaN in low_band at index ${i}`);
        }
    }
});

test('RANGE_FILTER batch default', () => {
    
    const close = new Float64Array(testData.close);
    
    
    const result = wasm.range_filter_batch_unified(close, {
        range_size: [2.618, 2.618, 0],
        range_period: [10, 10, 0],
        smooth_range: true,
        smooth_period: 27
    });
    
    
    assert(result.values, 'Should have values array');
    assert(result.combos, 'Should have combos array');
    assert(typeof result.rows === 'number', 'Should have rows count');
    assert(typeof result.cols === 'number', 'Should have cols count');
    
    
    assert.strictEqual(result.combos.length, 1);
    assert.strictEqual(result.rows, 1);
    assert.strictEqual(result.cols, close.length);
    
    
    assert.strictEqual(result.values.length, 3 * close.length);
    
    
    const filterValues = result.values.slice(0, close.length);
    
    
    const singleResult = wasm.range_filter_js(close, 2.618, 10, true, 27);
    for (let i = 0; i < close.length; i++) {
        if (isNaN(singleResult.filter[i]) && isNaN(filterValues[i])) {
            continue; 
        }
        assertClose(
            filterValues[i],
            singleResult.filter[i],
            1e-10,
            `Batch vs single mismatch at index ${i}`
        );
    }
});

test('RANGE_FILTER batch sweep', () => {
    
    
    const data = new Float64Array(50);
    for (let i = 0; i < 50; i++) {
        data[i] = Math.sin(i * 0.2) * 100 + 500;
    }
    
    const result = wasm.range_filter_batch_unified(data, {
        range_size: [2.0, 3.0, 0.5],  
        range_period: [10, 20, 5],    
        smooth_range: true,
        smooth_period: 15
    });
    
    
    assert.strictEqual(result.combos.length, 9);
    assert.strictEqual(result.rows, 9);
    assert.strictEqual(result.cols, 50);
    
    
    assert.strictEqual(result.values.length, 9 * 3 * 50);
    
    
    const expectedCombos = [];
    for (const rs of [2.0, 2.5, 3.0]) {
        for (const rp of [10, 15, 20]) {
            expectedCombos.push({ range_size: rs, range_period: rp });
        }
    }
    
    for (let i = 0; i < expectedCombos.length; i++) {
        assertClose(
            result.combos[i].range_size,
            expectedCombos[i].range_size,
            0.01,
            `Range size mismatch at combo ${i}`
        );
        assert.strictEqual(
            result.combos[i].range_period,
            expectedCombos[i].range_period,
            `Range period mismatch at combo ${i}`
        );
    }
});

test('RANGE_FILTER no poison', () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.range_filter_js(close);
    
    
    const poisonPatterns = [0x1111111111111111n, 0x2222222222222222n, 0x3333333333333333n];
    
    for (const [name, arr] of [
        ['filter', result.filter],
        ['high_band', result.high_band],
        ['low_band', result.low_band]
    ]) {
        for (let i = 0; i < arr.length; i++) {
            if (isNaN(arr[i])) continue;
            
            
            const buffer = new ArrayBuffer(8);
            const view = new DataView(buffer);
            view.setFloat64(0, arr[i], true);
            const bits = view.getBigUint64(0, true);
            
            assert(
                !poisonPatterns.includes(bits),
                `Poison pattern found in ${name} at index ${i}: 0x${bits.toString(16)}`
            );
        }
    }
});

test('RANGE_FILTER batch no poison', () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.range_filter_batch_unified(close, {
        range_size: [2.618, 2.618, 0],
        range_period: [10, 10, 0],
        smooth_range: true,
        smooth_period: 27
    });
    
    
    const poisonPatterns = [0x1111111111111111n, 0x2222222222222222n, 0x3333333333333333n];
    
    for (let i = 0; i < result.values.length; i++) {
        if (isNaN(result.values[i])) continue;
        
        const buffer = new ArrayBuffer(8);
        const view = new DataView(buffer);
        view.setFloat64(0, result.values[i], true);
        const bits = view.getBigUint64(0, true);
        
        assert(
            !poisonPatterns.includes(bits),
            `Poison pattern in values[${i}]: 0x${bits.toString(16)}`
        );
    }
});

test('RANGE_FILTER kernel parity', () => {
    
    const close = new Float64Array(testData.close.slice(0, 100)); 
    
    
    const resultAuto = wasm.range_filter_js(close);
    
    
    const resultSecond = wasm.range_filter_js(close);
    
    
    for (let i = 0; i < close.length; i++) {
        if (isNaN(resultAuto.filter[i]) && isNaN(resultSecond.filter[i])) {
            continue;
        }
        assertClose(
            resultAuto.filter[i],
            resultSecond.filter[i],
            1e-10,
            `Kernel consistency check failed at index ${i}`
        );
    }
});

test('RANGE_FILTER multi-output structure', () => {
    
    const close = new Float64Array(testData.close.slice(0, 100));
    
    const result = wasm.range_filter_js(close);
    
    
    assert(result.filter, 'Should have filter output');
    assert(result.high_band, 'Should have high_band output');
    assert(result.low_band, 'Should have low_band output');
    
    
    assert.strictEqual(result.filter.length, close.length);
    assert.strictEqual(result.high_band.length, close.length);
    assert.strictEqual(result.low_band.length, close.length);
    
    
    for (let i = 50; i < close.length; i++) { 
        if (!isNaN(result.filter[i])) {
            
            assert(
                result.high_band[i] >= result.filter[i] - 1e-6,
                `High band < filter at index ${i}`
            );
            
            assert(
                result.low_band[i] <= result.filter[i] + 1e-6,
                `Low band > filter at index ${i}`
            );
        }
    }
});

test('RANGE_FILTER batch metadata', () => {
    
    const close = new Float64Array(20);
    close.fill(100);
    
    const result = wasm.range_filter_batch_unified(close, {
        range_size: [2.0, 3.0, 1.0],      
        range_period: [10, 12, 2],        
        smooth_range: true,
        smooth_period: 15
    });
    
    
    assert.strictEqual(result.combos.length, 4);
    
    
    assert.strictEqual(result.combos[0].range_size, 2.0);
    assert.strictEqual(result.combos[0].range_period, 10);
    
    
    assert.strictEqual(result.combos[3].range_size, 3.0);
    assert.strictEqual(result.combos[3].range_period, 12);
});

test('RANGE_FILTER batch edge cases', () => {
    
    const close = new Float64Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);
    
    
    const singleBatch = wasm.range_filter_batch_unified(close, {
        range_size: [2.0, 2.0, 1],
        range_period: [5, 5, 1],
        smooth_range: true,
        smooth_period: 3
    });
    
    assert.strictEqual(singleBatch.combos.length, 1);
    assert.strictEqual(singleBatch.values.length, 3 * 10); 
    
    
    const largeBatch = wasm.range_filter_batch_unified(close, {
        range_size: [2.0, 2.5, 10], 
        range_period: [5, 5, 0],
        smooth_range: true,
        smooth_period: 3
    });
    
    
    assert.strictEqual(largeBatch.combos.length, 1);
    
    
    assert.throws(() => {
        wasm.range_filter_batch_unified(new Float64Array([]), {
            range_size: [2.618, 2.618, 0],
            range_period: [10, 10, 0],
            smooth_range: true,
            smooth_period: 27
        });
    }, /Input data slice is empty/);
});

test('RANGE_FILTER batch error handling', () => {
    const close = new Float64Array(testData.close.slice(0, 10));
    
    
    assert.throws(() => {
        wasm.range_filter_batch_unified(close, {
            range_size: [2.618, 2.618], 
            range_period: [10, 10, 0],
            smooth_range: true,
            smooth_period: 27
        });
    }, /Invalid config/);
    
    
    assert.throws(() => {
        wasm.range_filter_batch_unified(close, {
            range_size: [2.618, 2.618, 0],
            
            smooth_range: true,
            smooth_period: 27
        });
    }, /Invalid config/);
    
    
    assert.throws(() => {
        wasm.range_filter_batch_unified(close, {
            range_size: "invalid",
            range_period: [10, 10, 0],
            smooth_range: true,
            smooth_period: 27
        });
    }, /Invalid config/);
});

test('RANGE_FILTER batch full parameter sweep', () => {
    
    const close = new Float64Array(testData.close.slice(0, 50));
    
    const batchResult = wasm.range_filter_batch_unified(close, {
        range_size: [2.0, 2.5, 0.5],     
        range_period: [10, 11, 1],       
        smooth_range: false,
        smooth_period: 27
    });
    
    
    assert.strictEqual(batchResult.combos.length, 4);
    assert.strictEqual(batchResult.rows, 4);
    assert.strictEqual(batchResult.cols, 50);
    
    
    assert.strictEqual(batchResult.values.length, 4 * 3 * 50);
    
    
    for (let combo = 0; combo < batchResult.combos.length; combo++) {
        const range_size = batchResult.combos[combo].range_size;
        const range_period = batchResult.combos[combo].range_period;
        
        
        const baseIdx = combo * 3 * 50;
        const filterData = batchResult.values.slice(baseIdx, baseIdx + 50);
        
        
        let firstNonNan = -1;
        for (let i = 0; i < 50; i++) {
            if (!isNaN(filterData[i])) {
                firstNonNan = i;
                break;
            }
        }
        
        
        assert(firstNonNan >= 0 && firstNonNan < 50, 
               `No valid values found for combo ${combo} (size=${range_size}, period=${range_period})`);
    }
});

test('RANGE_FILTER SIMD128 consistency', () => {
    
    const testCases = [
        { size: 50 },
        { size: 100 },
        { size: 1000 }
    ];
    
    for (const testCase of testCases) {
        const data = new Float64Array(testCase.size);
        for (let i = 0; i < testCase.size; i++) {
            data[i] = Math.sin(i * 0.1) + Math.cos(i * 0.05);
        }
        
        const result = wasm.range_filter_js(data);
        
        
        assert.strictEqual(result.filter.length, data.length);
        assert.strictEqual(result.high_band.length, data.length);
        assert.strictEqual(result.low_band.length, data.length);
        
        
        let hasValues = false;
        for (let i = 10; i < result.filter.length; i++) {
            if (!isNaN(result.filter[i])) {
                hasValues = true;
                break;
            }
        }
        assert(hasValues, `No valid values found for size=${testCase.size}`);
    }
});

test.after(() => {
    console.log('RANGE_FILTER WASM tests completed');
});

File: tests/python/test_range_filter.py
"""
Python binding tests for RANGE_FILTER indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path

try:
    import my_project as ta_indicators
except ImportError:
    # If not in virtual environment, try to import from installed location
    try:
        import my_project as ta_indicators
    except ImportError:
        pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close, EXPECTED_OUTPUTS


class TestRangeFilter:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()
    
    def test_range_filter_accuracy(self, test_data):
        """Test RANGE_FILTER matches expected values from Rust tests - mirrors check_range_filter_accuracy"""
        close = test_data['close']
        
        # Use default parameters - returns tuple (filter, high_band, low_band)
        filter_values, high_band, low_band = ta_indicators.range_filter(close)
        
        assert len(filter_values) == len(close)
        assert len(high_band) == len(close)
        assert len(low_band) == len(close)
        
        # Test last 5 values against Rust references (exact same as Rust unit tests)
        expected_filter = [
            59_589.73987817684,
            59_589.73987817684,
            59_589.73987817684,
            59_589.73987817684,
            59_589.73987817684,
        ]

        expected_high = [
            60_935.63924911415,
            60_906.58379951138,
            60_874.2002431993,
            60_838.79850154794,
            60_810.879398758305,
        ]

        expected_low = [
            58_243.84050723953,
            58_272.8959568423,
            58_305.27951315438,
            58_340.68125480574,
            58_368.60035759538,
        ]

        # Match Rust test tolerance
        tolerance = 1e-10
        
        # Check Filter values
        last_5_filter = filter_values[-5:]
        for i, (val, exp) in enumerate(zip(last_5_filter, expected_filter)):
            diff = abs(val - exp)
            assert diff < tolerance, f"Filter[{i}] mismatch: expected {exp}, got {val} (diff: {diff})"
        
        # Check High Band values
        last_5_high = high_band[-5:]
        for i, (val, exp) in enumerate(zip(last_5_high, expected_high)):
            diff = abs(val - exp)
            assert diff < tolerance, f"High Band[{i}] mismatch: expected {exp}, got {val} (diff: {diff})"
        
        # Check Low Band values
        last_5_low = low_band[-5:]
        for i, (val, exp) in enumerate(zip(last_5_low, expected_low)):
            diff = abs(val - exp)
            assert diff < tolerance, f"Low Band[{i}] mismatch: expected {exp}, got {val} (diff: {diff})"
    
    def test_range_filter_default_candles(self, test_data):
        """Test RANGE_FILTER with default parameters - mirrors check_range_filter_default_candles"""
        close = test_data['close']
        
        filter_values, high_band, low_band = ta_indicators.range_filter(close)
        
        assert len(filter_values) == len(close)
        assert len(high_band) == len(close)
        assert len(low_band) == len(close)
    
    def test_range_filter_partial_params(self, test_data):
        """Test RANGE_FILTER with partial parameters - mirrors check_rf_partial_params"""
        close = test_data['close']
        
        # Test with only range_size specified
        filter_values, high_band, low_band = ta_indicators.range_filter(close, range_size=2.5)
        assert len(filter_values) == len(close)
        
        # Test with range_size and range_period
        filter_values, high_band, low_band = ta_indicators.range_filter(close, range_size=2.5, range_period=15)
        assert len(filter_values) == len(close)
        
        # Test with all parameters
        filter_values, high_band, low_band = ta_indicators.range_filter(
            close, 
            range_size=2.5, 
            range_period=15, 
            smooth_range=True,
            smooth_period=20
        )
        assert len(filter_values) == len(close)
    
    def test_range_filter_empty_input(self):
        """Test RANGE_FILTER fails with empty input - mirrors check_range_filter_empty_input"""
        empty = np.array([])
        
        with pytest.raises(ValueError, match="Input data slice is empty"):
            ta_indicators.range_filter(empty)
    
    def test_range_filter_all_nan(self):
        """Test RANGE_FILTER with all NaN values - mirrors check_range_filter_all_nan"""
        all_nan = np.full(100, np.nan)
        
        with pytest.raises(ValueError, match="All values are NaN"):
            ta_indicators.range_filter(all_nan)
    
    def test_range_filter_invalid_period(self):
        """Test RANGE_FILTER fails with invalid period - mirrors check_range_filter_invalid_period"""
        data = np.array([1.0, 2.0, 3.0])
        
        # Period exceeds data length
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.range_filter(data, range_period=10)
        
        # Zero period
        with pytest.raises(ValueError, match="Invalid period|Invalid range_period"):
            ta_indicators.range_filter(data, range_period=0)
    
    def test_range_filter_invalid_range_size(self):
        """Test RANGE_FILTER fails with invalid range_size - mirrors check_rf_invalid_range_size"""
        data = np.array([1.0, 2.0, 3.0, 4.0, 5.0])
        # Zero range_size
        with pytest.raises(ValueError, match="Invalid range_size"):
            ta_indicators.range_filter(data, range_size=0.0)
        # Negative range_size
        with pytest.raises(ValueError, match="Invalid range_size"):
            ta_indicators.range_filter(data, range_size=-1.0)
        # NaN range_size
        with pytest.raises(ValueError, match="Invalid range_size"):
            ta_indicators.range_filter(data, range_size=float('nan'))
    
    def test_range_filter_nan_handling(self, test_data):
        """Test RANGE_FILTER handles NaN values correctly - mirrors check_rf_nan_handling"""
        close = test_data['close']
        
        filter_values, high_band, low_band = ta_indicators.range_filter(close)
        
        # Check warmup period for NaNs
        # The warmup depends on the parameters but should be minimal
        # After index 50, should have no NaNs
        if len(filter_values) > 50:
            for i in range(50, len(filter_values)):
                assert not np.isnan(filter_values[i]), f"Found unexpected NaN in filter at index {i}"
                assert not np.isnan(high_band[i]), f"Found unexpected NaN in high_band at index {i}"
                assert not np.isnan(low_band[i]), f"Found unexpected NaN in low_band at index {i}"
    
    def test_range_filter_streaming(self, test_data):
        """Test RANGE_FILTER streaming matches batch calculation - mirrors check_range_filter_streaming"""
        close = test_data['close']
        
        # Batch calculation
        batch_filter, batch_high, batch_low = ta_indicators.range_filter(close)
        
        # Streaming calculation - use default parameters
        stream = ta_indicators.RangeFilterStream(
            range_size=2.618,
            range_period=14,
            smooth_range=True,
            smooth_period=27
        )
        stream_filter = []
        stream_high = []
        stream_low = []
        
        for price in close:
            result = stream.update(price)
            if result is not None:
                # Returns tuple (filter, high_band, low_band)
                stream_filter.append(result[0])
                stream_high.append(result[1])
                stream_low.append(result[2])
            else:
                stream_filter.append(np.nan)
                stream_high.append(np.nan)
                stream_low.append(np.nan)
        
        stream_filter = np.array(stream_filter)
        stream_high = np.array(stream_high)
        stream_low = np.array(stream_low)
        
        # Compare structure and basic properties (mirror Rust: no strict numeric equality)
        assert len(batch_filter) == len(stream_filter)
        # After warmup, ensure values are not NaN and bands are ordered
        start = 30
        for i in range(start, len(batch_filter)):
            assert not np.isnan(stream_filter[i]), f"Stream filter[{i}] is NaN"
            assert not np.isnan(stream_high[i]), f"Stream high[{i}] is NaN"
            assert not np.isnan(stream_low[i]), f"Stream low[{i}] is NaN"
            assert stream_high[i] >= stream_filter[i], f"High band < filter at index {i}"
            assert stream_filter[i] >= stream_low[i], f"Filter < low band at index {i}"
    
    def test_range_filter_batch_default(self, test_data):
        """Test RANGE_FILTER batch processing - mirrors check_range_filter_batch_default"""
        close = test_data['close']
        
        # Single parameter set - use start/end/step parameters
        result = ta_indicators.range_filter_batch(
            close,
            range_size_start=2.618,
            range_size_end=2.618,
            range_size_step=0.1,
            range_period_start=10,
            range_period_end=10,
            range_period_step=1,
            smooth_range=True,
            smooth_period=27
        )
        
        assert 'filter' in result
        assert 'high' in result
        assert 'low' in result
        assert 'range_sizes' in result
        assert 'range_periods' in result
        
        # Should have 1 combination
        assert result['rows'] == 1
        assert result['cols'] == len(close)
        assert result['filter'].shape[0] == 1
        assert result['filter'].shape[1] == len(close)
    
    def test_range_filter_batch_sweep(self, test_data):
        """Test RANGE_FILTER batch with parameter sweep - mirrors check_range_filter_batch_sweep"""
        # Use smaller dataset for speed
        data = np.sin(np.linspace(0, 10, 50)) * 100 + 500
        
        result = ta_indicators.range_filter_batch(
            data,
            range_size_start=2.0,
            range_size_end=3.0,
            range_size_step=0.5,  # 3 values: 2.0, 2.5, 3.0
            range_period_start=10,
            range_period_end=20,
            range_period_step=5,  # 3 values: 10, 15, 20
            smooth_range=True,
            smooth_period=15
        )
        
        # Should have 3 * 3 = 9 combinations
        assert result['rows'] == 9
        assert result['cols'] == len(data)
        assert result['filter'].shape[0] == 9
        assert result['filter'].shape[1] == len(data)
        assert result['high'].shape == result['filter'].shape
        assert result['low'].shape == result['filter'].shape
        
        # Verify parameters
        expected_sizes = []
        expected_periods = []
        for rs in [2.0, 2.5, 3.0]:
            for rp in [10, 15, 20]:
                expected_sizes.append(rs)
                expected_periods.append(rp)
        
        for i in range(9):
            assert abs(result['range_sizes'][i] - expected_sizes[i]) < 0.01
            assert result['range_periods'][i] == expected_periods[i]
    
    def test_range_filter_no_poison(self, test_data):
        """Test RANGE_FILTER doesn't leak poison patterns - mirrors check_range_filter_no_poison"""
        close = test_data['close']
        
        filter_values, high_band, low_band = ta_indicators.range_filter(close)
        
        # Check for poison patterns in all outputs
        poison_patterns = [0x1111111111111111, 0x2222222222222222, 0x3333333333333333]
        
        for name, arr in [('filter', filter_values), 
                          ('high_band', high_band), 
                          ('low_band', low_band)]:
            for i, val in enumerate(arr):
                if np.isnan(val):
                    continue
                # Convert to bits and check
                bits = np.array([val]).view(np.uint64)[0]
                assert bits not in poison_patterns, f"Poison pattern found in {name} at index {i}: 0x{bits:016X}"
    
    def test_range_filter_batch_no_poison(self, test_data):
        """Test batch RANGE_FILTER doesn't leak poison patterns - mirrors check_range_filter_batch_no_poison"""
        close = test_data['close']
        
        result = ta_indicators.range_filter_batch(
            close,
            range_size_start=2.618,
            range_size_end=2.618,
            range_size_step=0.1,
            range_period_start=10,
            range_period_end=10,
            range_period_step=1
        )
        
        # Check for poison patterns in all matrices
        poison_patterns = [0x1111111111111111, 0x2222222222222222, 0x3333333333333333]
        
        for matrix_name in ['filter', 'high', 'low']:
            matrix = result[matrix_name]
            for row in range(matrix.shape[0]):
                for col in range(matrix.shape[1]):
                    val = matrix[row, col]
                    if np.isnan(val):
                        continue
                    bits = np.array([val]).view(np.uint64)[0]
                    assert bits not in poison_patterns, f"Poison pattern in {matrix_name}[{row},{col}]: 0x{bits:016X}"
    
    def test_range_filter_kernel_parity(self, test_data):
        """Test different kernels produce similar results - mirrors check_range_filter_kernel_parity"""
        close = test_data['close'][:100]  # Use smaller dataset
        
        # Get result with default (auto) kernel
        filter_auto, high_auto, low_auto = ta_indicators.range_filter(close)
        
        # The kernels should produce very similar results
        # We can't control kernel selection from Python, but we can verify consistency
        filter_second, high_second, low_second = ta_indicators.range_filter(close)
        
        # Results should be identical when using same kernel
        for i in range(len(close)):
            if np.isnan(filter_auto[i]) and np.isnan(filter_second[i]):
                continue
            assert_close(
                filter_auto[i],
                filter_second[i],
                rtol=1e-10,
                msg=f"Kernel consistency check failed at index {i}"
            )
    
    def test_range_filter_multi_output_structure(self, test_data):
        """Test Range Filter returns all three outputs correctly"""
        close = test_data['close'][:100]
        
        filter_values, high_band, low_band = ta_indicators.range_filter(close)
        
        # Check all have same length as input
        assert len(filter_values) == len(close)
        assert len(high_band) == len(close)
        assert len(low_band) == len(close)
        
        # Check relationship between outputs (high > filter > low where not NaN)
        for i in range(50, len(close)):  # Skip warmup
            if not np.isnan(filter_values[i]):
                # High band should be >= filter
                assert high_band[i] >= filter_values[i] - 1e-6, \
                    f"High band < filter at index {i}"
                # Low band should be <= filter  
                assert low_band[i] <= filter_values[i] + 1e-6, \
                    f"Low band > filter at index {i}"


if __name__ == '__main__':
    pytest.main([__file__, '-v'])
